src/Generator/TypeGenerator.php

```diff
@@ -136,30 +129,14 @@ public static function fromTypeString(string $type): self
             ));
         }
 
-        foreach ($types as $index => $atomicType) {
-            $otherTypes = array_diff_key($types, array_flip([$index]));
-
-            assert([] !== $otherTypes, 'There are always 2 or more types in a union type');
-
-            if ($isIntersection) {
-                $atomicType->assertCanIntersectWith($otherTypes);
-            } else {
-                $atomicType->assertCanUnionWith($otherTypes);
-            }
-        }
-
-        return new self($types, $nullable, $isIntersection);
+        return new self(CompositeType::fromString($trimmedNullable));
     }
 
-    /**
-     * @param AtomicType[]                     $types
-     * @psalm-param non-empty-list<AtomicType> $types
-     */
-    private function __construct(array $types, bool $nullable, bool $isIntersection)
+    private function __construct(private readonly TypeInterface $type, private readonly bool $nullable = false)
     {
-        $this->types          = $types;
-        $this->nullable       = $nullable;
-        $this->isIntersection = $isIntersection;
+        if ($nullable && $type instanceof AtomicType) {
+            $type->assertCanBeStandaloneNullable();
+        }
```

Comment:
Should I keep it here or better move back to `fromString()`?

Reply:
Works here too :+1: